### DISCUSSION TITLE
async await이 뭐지?

### DISCUSSION THREAD

**Discussion by Yeggggi** [DeveloperAcademy-POSTECH/.github#34] [Idea]
며칠전에 개발을 하다보면 async - await 를 분명 많이 쓰는데 뭔지모르고 그냥 쓰고 있어서 혹시 설명해줄 수 있는지 질문을 받았습니다. 사실 여전히 관련해서 봐야할 것이 너무 많고 설명해줄만큼 정확히는 모르지만 
sync, async, await 개념만 조금 이야기 해보려고합니다!
&nbsp;
&nbsp;

sync 와 async가 무엇일까?
-> sync(동기) / async (비동기) 는 프로세스의 작업 순서 보장에 대한 개념입니다.

**동기(Synchronous)**

- 현재 작업의 처리가 끝나는 시간에 맞춰 동시에 다음 작업을 요청
- 함수를 호출한 곳에서 호출된 함수가 결과를 반환할 때 까지 대기
- 작업 완료 여부를 계속해서 확인
&nbsp;

**비동기(Asynchronous)**

- 현재 작업의 응답이 끝나지 않은 상태에서 다음 작업을 요청
- 함수를 호출한 곳에서 완료 여부를 신경쓰지 않고, 호출된 곳에서 작업이 끝나면 **callback으로 결과를 전달**
- 작업 완료 여부를 확인하지 않음

비동기 처리는 결과를 즉시 받을 필요가 없거나 오래걸리는 작업을 위해 사용됩니다. 

&nbsp;

햄버거 가게를 예로 들어 보겠습니다. 손님이 햄버거를 주문하면 
**sync(동기)** 는 햄버거가 완성되었는지 계속 물어보면서 순서대로 대기합니다.
**async(비동기)** 는 햄버거가 완성되었는지 신경쓰지 않고 있다가 진동벨이 울리면 전달받습니다.

&nbsp;
&nbsp;

 **await**

await는 async(비동기) 함수를 sync(동기)처럼 쓸 수 있도록 해줍니다. 

그리고 await는 async 함수 안에서만 동작하기 때문에 async 함수 내부에서만 사용해야 합니다. 그래서 await를 사용해야 한다면 await를 감싸는 함수에 async를 넣어 주어야 합니다.

아래처럼 async throws, try await의 형태로 함께 사용하는 것을 많이 볼 수 있는데 throws - try는 에러 반환을 위해 필요합니다. 

```swift
 func reportServiceIsReconnecting() async throws {
        try await channelManager.setServiceStatus(.connecting, channelUUID: channelUUID)
    }
```

&nbsp;
결론적으로 함수들의 실행순서가 중요하거나, 해당 함수의 실행 결과값이 중요한데 비동기(async)인 경우에 await을 함께 써주어서 처리완료를 기다렸다가 결과값을 얻도록 합니다. 
&nbsp;
&nbsp;
또한 async await는 비동기 함수를 동기처럼 쓸 수 있도록하지만 동기처럼 작업이 완료되어야만 쓰레드를 반환하는 것이 아니라 시스템에게 제어권을 맡겨서 작업 상황에 따라 비동기함수를 일시 중단 할 수 있도록 합니다. 또한 이 일시 중단될 수 있는 위치를 지정하기위해 await 키워드를 사용합니다. 

&nbsp;
&nbsp;
&nbsp;
혹시 관련해서 수정이 필요하거나 첨언해줄 것이 있으시다면 달아주시면 너무너무 감사하겠습니다☺️
&nbsp;

**Comment by moonkey48** [comment]
> 좋은 글 감사합니다! :) 동기 비동기에 관한 리소스를 제작하고 있었어서 추가로 적어둡니다!ㅎㅎ 올려주신 내용은 Swift에서 비동기 방식의 한 종류인 async-await 방식에 대해 잘 설명해주신 것 같아요!ㅎㅎ 추가로 어떤 비동기장식이 가능하고, 사용될 수 있는지 알아보시는 것도 다양한 방법의 비동기 코드를 이해하고 적용할 수 있을 것 같아서 공부해보시면 좋을 것 같습니다  😄

## 동기와 비동기의 개념

### 동기(Synchronous)

> 동기는 어떤 작업, 테스크들이 들어왔을 때(ex. 화면 이동, 버튼을 클릭 등), 해당 테스크들을 순차적으로 진행합니다. A, B, C 세가지의 작업이 들어왔다면, 작업을 `A 처리 → A 완료 → B 처리 → B 완료 → C 처리 → C 완료`의 순서로 진행되게 됩니다. 이것이 동기의 개념입니다.
> 

> 동기 방식은 현재 진행중인 테스크를 처리해야 다음 연산을 진행하기 때문에 쉽게 예측이 가능합니다. 하지만 특정 테스크 작업 시간이 오래 걸리는 경우, 처리중이지만 멈춰있는 것처럼 보일 수 있습니다.
> 

### 비동기(Asynchoronomous)

> 비동기는 동기와 다르게, 테스크들을 동시에, 병렬적으로 처리하게 됩니다. 위에 예시와 같이 A, B, C의 작업이 들어왔다면 A `처리(0.001초) → B처리(0.001초) → C처리(0.001초)`와 같이 여러개의 작업을 번갈아 가며 처리를 하게 됩니다. 동시간대의 여러개의 테스크를 수행하는 것이 동기와 다른 비동기의 특징입니다.
> 

> 네트워크 호출,  큰 연산처리 등의 시간이 많이 드는 테스크를 수행할 때 용이합니다. 예를 들어 로그인을 요청을 보내고, 요청에 대한 답을 기다리는 시간이 1초 이상 걸리는 경우 비동기를 통해 해당 작업 중에는 ProgressView를 보여줘서 처리가 진행중인 사실을 보여줄 수 있습니다.
> 

## 언제 동기와 비동기가 중요한가요?

1. **성능 향상**
    
    비동기 프로그래밍은 동시에 여러 작업을 수행할 수 있으므로, 하나의 작업이 완료되는 동안 다른 작업을 진행할 수 있습니다. 이로 인해 전체 작업의 처리 시간을 단축할 수 있으며 사용자 경험을 향상시킬 수 있습니다.
    
2. **응답성 향상**
    
    긴 작업을 수행하는 동안 UI가 차단되지 않으므로, 사용자는 다른 작업을 계속 수행할 수 있습니다. 이는 특히 웹 및 모바일 애플리케이션에서 중요하며, 사용자와의 상호 작용을 유지하면서 백그라운드에서 복잡한 작업을 수행할 수 있습니다.
    
3. **자원 최적화**
    
    비동기 프로그래밍을 사용하면 시스템 자원을 보다 효율적으로 사용할 수 있습니다. 예를 들어, 네트워크 요청을 기다리는 동안 CPU는 다른 계산을 수행할 수 있으며, 이로 인해 전체 시스템의 효율성이 향상됩니다.
    
4. **유연성과 확장성**
    
    비동기 코드는 모듈화 및 재사용이 용이하며, 다양한 환경과 상황에서 확장성 있게 동작할 수 있습니다.
    

## Swift에서의 동기와 비동기

### 동기

> 사실 Swift에서 여러분들이 작성한 코드는 동기로 작동합니다. 아래 예시처럼 위에서 아래로, 순차적으로 함수가 호출되고 진행중인 함수가 종료되기 전까지는 아래에 있는 다른 함수가 호출되지 않습니다.
> 

```swift
func sayHello() {
	print("Hello")
}
func sayName(_ name: String) {
	print("\(name)!!")
}
func sayGoodbye() {
	print("Have a good day! :)")
}

sayHello()
sayName("Austin")
sayGoodbye()

/*
Hello
Austin!!
Have a good day! :)
*/
```

### 비동기
- Async-Await
- ? 
- ?



### 참고
https://devhyeon0312.tistory.com/13
